src/onegov/search/indexer.py

```diff
@@ -410,45 +412,139 @@ def index(
                     if not k.startswith('es_')}
-                _id = task['id']
-                content.append(
-                    {'language': language, 'data': data, '_id': _id})
+                # the search makes use of 'es_public' and 'es_tags' fields
+                for k in ('es_public', 'es_tags', 'es_last_change'):
+                    if k in task['properties']:
+                        data[k] = task['properties'][k]
+                _owner_id: int | UUID | str = task['owner_id']
+                _owner_type = task['owner_type']  # class name
+
+                params.append({
+                    '_language': language,
+                    '_data': data,
+                    '_owner_id_int':
+                        _owner_id if isinstance(_owner_id, int) else None,
+                    '_owner_id_uuid':
+                        _owner_id if isinstance(_owner_id, UUID) else None,
+                    '_owner_id_str':
+                        _owner_id if isinstance(_owner_id, str) else None,
+                    '_owner_type': _owner_type,
+                    **{f'_{k}': v for k, v in data.items()}
+                })
+
+            weighted_vector = [
+                func.setweight(
+                    func.to_tsvector(
+                        sqlalchemy.bindparam('_language',
+                                             type_=sqlalchemy.String),
+                        sqlalchemy.bindparam(f'_{field}',
+                                             type_=sqlalchemy.String)
+                    ),
+                    weight
+                )
+                for field, weight in zip(
+                    tasks[0]['properties'].keys(),
+                    chain('A', repeat('D')))
+                if not field.startswith('es_')  # TODO: rename to fts_
+            ]
+
+            combined_vector = weighted_vector[0]
+            for vector in weighted_vector[1:]:
+                combined_vector = combined_vector.op('||')(vector)
 
             schema = tasks[0]['schema']
-            tablename = tasks[0]['tablename']
-            id_key = tasks[0]['id_key']
-            table = sqlalchemy.table(
-                tablename,
-                (id_col := sqlalchemy
-                    .column(id_key)),  # type: ignore[var-annotated]
-                sqlalchemy.column(self.TEXT_SEARCH_COLUMN_NAME),
-                schema=schema  # type: ignore
-            )
-            tsvector_expr = sqlalchemy.text(
-                'to_tsvector(:language, :data)').bindparams(
-                sqlalchemy.bindparam('language', type_=sqlalchemy.String),
-                sqlalchemy.bindparam('data', type_=sqlalchemy.JSON)
-            )
+
             stmt = (
-                sqlalchemy.update(table)
-                .where(id_col == sqlalchemy.bindparam('_id'))
-                .values({self.TEXT_SEARCH_COLUMN_NAME: tsvector_expr})
+                sqlalchemy.insert(SearchIndex.__table__)
```

Comment:
Postgres supports upserting, so rather than `DELETE` +`INSERT` we should run a single upsert query (`ON CONFLICT "unique_constraint_name" DO UPDATE SET fts_index = EXCLUDED.fts_index ...`). You will need to add a combined unique index on the four owner columns though. So Postgres detects that you're trying to insert a row that's already in the index.